clippy_lints/src/strings.rs

```diff
@@ -451,3 +451,64 @@ impl<'tcx> LateLintPass<'tcx> for StringToString {
         }
     }
 }
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Warns about calling `str::trim` (or variants) before `str::split_whitespace`.
+    ///
+    /// ### Why is this bad?
+    /// `split_whitespace` already ignores leading and trailing whitespace.
+    ///
+    /// ### Example
+    /// ```rust
+    /// " A B C ".trim().split_whitespace();
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// " A B C ".split_whitespace();
+    /// ```
+    #[clippy::version = "1.62.0"]
+    pub TRIM_SPLIT_WHITESPACE,
+    style,
+    "using `str::trim()` or alike before `str::split_whitespace`"
+}
+declare_lint_pass!(TrimSplitWhitespace => [TRIM_SPLIT_WHITESPACE]);
+
+impl<'tcx> LateLintPass<'tcx> for TrimSplitWhitespace {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &Expr<'_>) {
+        if_chain! {
+            if let ExprKind::MethodCall(path, [recv], split_ws_span) = expr.kind;
+            if path.ident.name == sym!(split_whitespace);
+            if let ExprKind::MethodCall(path, [recv], trim_span) = recv.kind;
+            if let trim_fn_name @ ("trim" | "trim_start" | "trim_end") = path.ident.name.as_str();
+            let recv_ty = cx.typeck_results().expr_ty(recv).peel_refs();
+            if recv_ty.is_str() || implements_deref_str(cx, recv_ty);
+            then {
+                span_lint_and_sugg(
+                    cx,
+                    TRIM_SPLIT_WHITESPACE,
+                    split_ws_span.with_lo(trim_span.lo()),
+                    format!("found call to `str::{}` before `str::split_whitespace`", trim_fn_name).as_str(),
+                    format!("remove `{}()`", trim_fn_name).as_str(),
+                    "split_whitespace()".to_string(),
+                    Applicability::MachineApplicable,
+                );
+            }
+        }
+    }
+}
+
+/// does ty implement Deref<Target=str>?
+fn implements_deref_str<'t>(cx: &LateContext<'t>, ty: Ty<'t>) -> bool {
```

Comment:
With diag items, not sure if this is even necessary, because you only check for the functions implemented for `str`.